File: rig/fraud.py
"""Page fraud proofs — the dispute game (Sharding Road, Phase 1).

A `PageFraudProof` convicts a block of committing a WRONG post-state for one
page, verifiable from first principles by a node that holds NOTHING but the
two block headers. This is the mechanism that lets a future paged validator
trust pages it does not hold: any single honest holder can convict a bad
aggregation with one bounded proof.

The proof is a plain dict (JSON-friendly — it is golden-vectored and travels
over gossip):

    header          accused block's header fields
    parent_header   its parent's header fields
    parent_model_json  the parent ModelState's CANONICAL json — bound by
                    sha256(json) == parent_header.model_root, so page spans
                    and dim need no trusted context
    committed_leaves   hex page-leaf hashes the accused block committed —
                    bound by fold(leaves) == header.state_root (the accused
                    root itself authenticates them; post-v5 blocks carry them
                    in the body, but the verifier never cares where they came
                    from)
    page_id         the disputed page
    txids           ALL txids in the block (recomputes header.txset_root, so
                    the claimant set is provably complete)
    txs             all tx field-dicts (small; sigs verified)
    bodies          txid -> sparse body {n, idx: [u32], val: [i64]} for every
                    tx CLAIMING the page — delta_hash binds each to its tx
                    (the hash is over the dense byte image; sparse is
                    transport)
    parent_page     the parent state's page values (ints)
    parent_path     merkle inclusion proof of parent_page in
                    parent_header.state_root: [(side, hex)]

verify() returns (fraud_proven, reason). Reasons distinguish an INVALID proof
(never act on it) from a VALID one (the block is provably fraudulent).
Scope v1: aggregation fraud on existing pages of non-growth blocks; growth
and fold disputes are checkable from the (tiny) ModelState alone and get
their own game later.
"""

import hashlib
import json

import numpy as np

from rig import merkle
from rig.blockchain import Header
from rig.chain import trimmed_mean_int
from rig.crypto import BackpropTx, delta_hash


def _sha(b: bytes) -> str:
    return hashlib.sha256(b).hexdigest()


def _dense(body: dict, dim: int) -> np.ndarray | None:
    idx = body.get("idx", [])
    val = body.get("val", [])
    if body.get("n") != dim or len(idx) != len(val):
        return None
    out = np.zeros(dim, dtype=np.int64)
    last = -1
    for i, v in zip(idx, val):
        if not (0 <= int(i) < dim) or int(i) <= last:
            return None  # coords must be sorted unique in range
        last = int(i)
        out[int(i)] = int(v)
    return out


def build(header: Header, parent_header: Header, parent_model_json: str,
          committed_leaves: list[str], page_id: int, txs: list[BackpropTx],
          bodies_dense: dict[str, np.ndarray],
          parent_state: np.ndarray) -> dict:
    """Assemble a proof from a prover's local data (a full validator that just
    rejected the block). Sparse-encodes only the claimant bodies."""
    span_pages = json.loads(parent_model_json)["pages"]
    s, e = span_pages[page_id][0], span_pages[page_id][1]
    dim = span_pages[-1][1]
    proof_bodies = {}
    for t in txs:
        if page_id in t.canonical_pages():
            d = bodies_dense[t.txid()]
            nz = np.nonzero(d)[0]
            proof_bodies[t.txid()] = {
                "n": dim,
                "idx": [int(i) for i in nz],
                "val": [int(d[i]) for i in nz],
            }
    parent_leaves = [np.asarray(parent_state[p[0]:p[1]], dtype=np.int64)
                     for p in span_pages]
    levels = merkle.build([pl.tobytes() for pl in parent_leaves])
    return {
        "header": dict(header.__dict__),
        "parent_header": dict(parent_header.__dict__),
        "parent_model_json": parent_model_json,
        "committed_leaves": committed_leaves,
        "page_id": page_id,
        "txids": sorted(t.txid() for t in txs),
        "txs": [{"miner": t.miner, "base_height": t.base_height,
                 "delta_hash": t.delta_hash, "da_pointer": t.da_pointer,
                 "bond": t.bond, "pages": t.canonical_pages(),
                 "data_refs": t.canonical_refs(), "sig_hex": t.sig.hex()}
                for t in txs],
        "bodies": proof_bodies,
        "parent_page": [int(x) for x in parent_state[s:e]],
        "parent_path": [(side, sib.hex())
                        for side, sib in merkle.proof(levels, page_id)],
    }


def verify(proof: dict) -> tuple[bool, str]:
    """First-principles verification. (True, ...) == the block is fraudulent."""
    try:
        h = Header(**proof["header"])
        ph = Header(**proof["parent_header"])
    except TypeError:
        return False, "invalid: malformed headers"
    if h.prev_hash != ph.block_hash():
        return False, "invalid: header does not extend parent"
    # parent ModelState binds spans + dim through model_root
    pmj = proof["parent_model_json"]
    if _sha(pmj.encode()) != ph.model_root:
        return False, "invalid: parent model json does not match model_root"
    pages = json.loads(pmj)["pages"]
    pid = proof["page_id"]
    if not (0 <= pid < len(pages)):
        return False, "invalid: page id out of range"
    s, e = pages[pid][0], pages[pid][1]
    dim = pages[-1][1]
    # committed leaves bind through the accused state_root itself
    leaves = proof["committed_leaves"]
    if len(leaves) != len(pages):
        return False, "invalid: growth blocks are not disputable in fraud v1"
    level = [bytes.fromhex(x) for x in leaves]
    while len(level) > 1:
        level = [merkle._node_hash(level[i],
                                   level[i + 1] if i + 1 < len(level)
                                   else level[i])
                 for i in range(0, len(level), 2)]
    if level[0].hex() != h.state_root:
        return False, "invalid: committed leaves do not fold to state_root"
    # tx set completeness through txset_root
    txids = proof["txids"]
    if _sha("|".join(sorted(txids)).encode()) != h.txset_root:
        return False, "invalid: txids do not reproduce txset_root"
    if len(proof["txs"]) != len(txids) or len(txids) != h.n_txs:
        return False, "invalid: tx list incomplete"
    txs = []
    for t in proof["txs"]:
        tx = BackpropTx(miner=t["miner"], base_height=t["base_height"],
                        delta_hash=t["delta_hash"], da_pointer=t["da_pointer"],
                        bond=t["bond"], pages=list(t["pages"]),
                        data_refs=list(t["data_refs"]))
        tx.sig = bytes.fromhex(t["sig_hex"])
        if tx.txid() not in txids:
            return False, "invalid: tx not in committed set"
        if not tx.verify():
            return False, "invalid: bad tx signature"
        if tx.base_height != ph.height:
            return False, "invalid: tx base height mismatch"
        txs.append(tx)
    # claimant bodies bind through delta_hash (over the dense byte image)
    claim_cols = []
    for tx in txs:
        if pid not in tx.canonical_pages():
            continue
        body = proof["bodies"].get(tx.txid())
        if body is None:
            return False, "invalid: missing claimant body"
        d = _dense(body, dim)
        if d is None:
            return False, "invalid: malformed body"
        if delta_hash(d.tobytes()) != tx.delta_hash:
            return False, "invalid: body does not match delta_hash"
        claim_cols.append(d[s:e])
    # parent page binds through parent state_root
    parent_page = np.asarray(proof["parent_page"], dtype=np.int64)
    if len(parent_page) != e - s:
        return False, "invalid: parent page span mismatch"
    path = [(side, bytes.fromhex(sib)) for side, sib in proof["parent_path"]]
    if not merkle.verify(parent_page.tobytes(), pid, path,
                         bytes.fromhex(ph.state_root)):
        return False, "invalid: parent page not in parent state_root"
    # THE RECOMPUTATION — the exact consensus rule, one page wide
    if claim_cols:
        new_page = parent_page + trimmed_mean_int(claim_cols)
    else:
        new_page = parent_page
    recomputed = merkle.leaf_hash(new_page.astype(np.int64).tobytes())
    committed = bytes.fromhex(leaves[pid])
    if recomputed == committed:
        return False, "invalid: page aggregates correctly — no fraud"
    return True, (f"FRAUD: page {pid} committed leaf {committed.hex()[:12]}… "
                  f"but honest aggregation gives {recomputed.hex()[:12]}…")

File: node/core/src/fraud.rs
//! Page fraud-proof verifier (Sharding Road, Phase 1) — the Rust side of the
//! dispute game, bit-exact with `rig/fraud.py::verify`.
//!
//! A proof convicts a block of committing a wrong post-state for ONE page,
//! checkable by a node that holds only the two block headers. The verdict
//! must match the reference exactly: acting on a false conviction, or missing
//! a real one, both fork the chain in a sharded world.

use crate::merkle;
use crate::int64_bytes;
use crate::{delta_hash, trimmed_mean_scalar, BackpropTx, Header};
use sha2::{Digest, Sha256};

fn sha_hex(b: &[u8]) -> String {
    hex::encode(Sha256::digest(b))
}

/// A claimant's sparse body inside a proof.
pub struct Body {
    pub n: usize,
    pub idx: Vec<u64>,
    pub val: Vec<i64>,
}

/// A tx record carried by a proof (core has no serde derive on BackpropTx;
/// the net/test layer parses JSON and hands these in).
pub struct TxRec {
    pub tx: BackpropTx,
}

/// A page fraud proof, fully parsed. The net layer and the golden test build
/// this from JSON; the verifier is pure logic over owned data.
pub struct PageFraudProof {
    pub header: Header,
    pub parent_header: Header,
    pub parent_model_json: String,
    pub committed_leaves: Vec<String>,
    pub page_id: usize,
    pub txids: Vec<String>,
    pub txs: Vec<TxRec>,
    pub bodies: std::collections::HashMap<String, Body>,
    pub parent_page: Vec<i64>,
    pub parent_path: Vec<(bool, String)>,
}

/// (fraud_proven, reason). `true` == the accused block is provably fraudulent.
/// A `false` with an "invalid:" reason means the PROOF is malformed and must
/// never be acted on.
pub fn verify(p: &PageFraudProof) -> (bool, String) {
    let inv = |m: &str| (false, format!("invalid: {m}"));

    if p.header.prev_hash != p.parent_header.block_hash() {
        return inv("header does not extend parent");
    }
    if sha_hex(p.parent_model_json.as_bytes()) != p.parent_header.model_root {
        return inv("parent model json does not match model_root");
    }
    let mj: serde_json::Value = match serde_json::from_str(&p.parent_model_json) {
        Ok(m) => m,
        Err(_) => return inv("unparseable parent model json"),
    };
    let pages = match mj.get("pages").and_then(|x| x.as_array()) {
        Some(a) if !a.is_empty() => a,
        _ => return inv("parent model has no pages"),
    };
    let span = |i: usize| -> (usize, usize) {
        let r = pages[i].as_array().unwrap();
        (r[0].as_u64().unwrap() as usize, r[1].as_u64().unwrap() as usize)
    };
    let pid = p.page_id;
    if pid >= pages.len() {
        return inv("page id out of range");
    }
    let (s, e) = span(pid);
    let dim = span(pages.len() - 1).1;

    // committed leaves bind through the accused state_root itself
    if p.committed_leaves.len() != pages.len() {
        return inv("growth blocks are not disputable in fraud v1");
    }
    let mut level: Vec<[u8; 32]> = Vec::with_capacity(p.committed_leaves.len());
    for h in &p.committed_leaves {
        match hex::decode(h).ok().and_then(|b| b.try_into().ok()) {
            Some(a) => level.push(a),
            None => return inv("malformed committed leaf"),
        }
    }
    if hex::encode(merkle::root_from_hashes(level.clone())) != p.header.state_root {
        return inv("committed leaves do not fold to state_root");
    }

    // tx-set completeness through txset_root
    let mut sorted_ids = p.txids.clone();
    sorted_ids.sort();
    if sha_hex(sorted_ids.join("|").as_bytes()) != p.header.txset_root {
        return inv("txids do not reproduce txset_root");
    }
    if p.txs.len() != p.txids.len() || p.txids.len() as u64 != p.header.n_txs {
        return inv("tx list incomplete");
    }

    // claimant bodies bind through delta_hash; recompute the page column set
    let mut claim_cols: Vec<Vec<i64>> = Vec::new();
    for rec in &p.txs {
        let tx = &rec.tx;
        if !p.txids.contains(&tx.txid()) {
            return inv("tx not in committed set");
        }
        if !tx.verify() {
            return inv("bad tx signature");
        }
        if tx.base_height != p.parent_header.height {
            return inv("tx base height mismatch");
        }
        if !tx.canonical_pages().contains(&(pid as u32)) {
            continue;
        }
        let body = match p.bodies.get(&tx.txid()) {
            Some(b) => b,
            None => return inv("missing claimant body"),
        };
        if body.n != dim || body.idx.len() != body.val.len() {
            return inv("malformed body");
        }
        // dense image (for the hash) + the page column
        let mut dense = vec![0i64; dim];
        let mut last: i64 = -1;
        for (i, v) in body.idx.iter().zip(&body.val) {
            let i = *i as usize;
            if i >= dim || i as i64 <= last {
                return inv("body coords not sorted-unique in range");
            }
            last = i as i64;
            dense[i] = *v;
        }
        if delta_hash(&int64_bytes(&dense)) != tx.delta_hash {
            return inv("body does not match delta_hash");
        }
        claim_cols.push(dense[s..e].to_vec());
    }

    // parent page binds through the parent state_root
    if p.parent_page.len() != e - s {
        return inv("parent page span mismatch");
    }
    let mut path: Vec<(bool, [u8; 32])> = Vec::with_capacity(p.parent_path.len());
    for (side, sib) in &p.parent_path {
        match hex::decode(sib).ok().and_then(|b| b.try_into().ok()) {
            Some(a) => path.push((*side, a)),
            None => return inv("malformed branch hash"),
        }
    }
    let parent_leaf = merkle::leaf_hash(&int64_bytes(&p.parent_page));
    let root: [u8; 32] = match hex::decode(&p.parent_header.state_root)
        .ok().and_then(|b| b.try_into().ok())
    {
        Some(a) => a,
        None => return inv("malformed parent state_root"),
    };
    if !merkle::verify_leaf(parent_leaf, &path, &root) {
        return inv("parent page not in parent state_root");
    }

    // THE RECOMPUTATION — the exact per-page consensus rule, one page wide
    let width = e - s;
    let mut new_page = p.parent_page.clone();
    if !claim_cols.is_empty() {
        for j in 0..width {
            let mut col: Vec<i64> = claim_cols.iter().map(|c| c[j]).collect();
            new_page[j] = new_page[j].wrapping_add(trimmed_mean_scalar(&mut col, 0.2));
        }
    }
    let recomputed = merkle::leaf_hash(&int64_bytes(&new_page));
    let committed = level[pid];
    if recomputed == committed {
        return (false, "invalid: page aggregates correctly — no fraud".into());
    }
    (true, format!("FRAUD: page {pid} committed {}… but honest {}…",
                   &hex::encode(committed)[..12], &hex::encode(recomputed)[..12]))
}
